packages/core/admin/admin/src/features/BackButton.tsx

```diff
@@ -0,0 +1,231 @@
+    dispatch({
+      type: 'PUSH_STATE',
+      payload: typeof path === 'string' ? { to: path, search: '' } : path,
+    });
+  }, []);
+
+  const goBack: HistoryContextValue['goBack'] = React.useCallback(() => {
+    /**
+     * Perform the browser back action
+     * dispatch the goBack action to keep redux in sync
+     * and set the ref to avoid an infinite loop and incorrect state pushing
+     */
+    navigate(-1);
+    dispatch({ type: 'GO_BACK' });
+    isGoingBack.current = true;
+  }, [navigate]);
+
+  /**
+   * This is a semi-listener pattern to keep the `canGoBack` state in sync.
+   */
+  const prevIndex = React.useRef(state.currentLocationIndex);
+  React.useEffect(() => {
+    if (state.currentLocationIndex !== prevIndex.current) {
+      dispatch({
+        type: 'SET_ABILITY_TO_NAVIGATE',
+        payload: state.currentLocationIndex > 1 && state.history.length > 1,
+      });
+      prevIndex.current = state.currentLocationIndex;
+    }
+  }, [prevIndex, state.currentLocationIndex, state.history.length]);
+
+  /**
+   * This effect is responsible for pushing the new state to the history
+   * when the user navigates to a new location assuming they're not going back.
+   */
+  React.useLayoutEffect(() => {
+    if (isGoingBack.current) {
+      isGoingBack.current = false;
+      return;
+    } else {
+      // this should only occur on link movements, not back/forward clicks
+      dispatch({
+        type: 'PUSH_STATE',
+        payload: { to: location.pathname, search: location.search },
+      });
+    }
+  }, [dispatch, location.pathname, location.search]);
+
+  return (
+    <Provider pushState={pushState} goBack={goBack} {...state}>
+      {children}
+    </Provider>
+  );
+};
+
+type HistoryActions =
+  | {
+      type: 'PUSH_STATE';
+      payload: {
+        to: string;
+        search: string;
+      };
+    }
+  | {
+      type: 'GO_BACK';
+    }
+  | {
+      type: 'SET_ABILITY_TO_NAVIGATE';
+      payload: boolean;
+    };
+
+const reducer = (state: HistoryState, action: HistoryActions) =>
+  produce(state, (draft) => {
+    switch (action.type) {
+      case 'PUSH_STATE': {
+        const path = `${action.payload.to}${action.payload.search}`;
+        if (state.currentLocationIndex === state.history.length) {
+          // add the new place
+          draft.history = [...state.history, path];
+        } else {
+          // delete all the history after the current place and then add the new place
+          draft.history = [...state.history.slice(0, state.currentLocationIndex), path];
+        }
+
+        draft.currentLocation = path;
+        draft.currentLocationIndex += 1;
+
+        break;
+      }
+      case 'GO_BACK': {
+        const newIndex = state.currentLocationIndex - 1;
+
+        draft.currentLocation = state.history[newIndex - 1];
+        draft.currentLocationIndex = newIndex;
+        break;
+      }
+      case 'SET_ABILITY_TO_NAVIGATE': {
```

Comment:
```suggestion
      case 'SET_CAN_GO_BACK': {
```

idm but does this make more sense?

Reply:
yeah this was taken from something else where i allowed forward too...